system/loongarch/cpuinfo.c

```diff
@@ -0,0 +1,281 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (C) 2020-2022 Martin Whitaker.
+// Copyright (C) 2004-2023 Sam Demeulemeester
+// Copyright (C) 2024 Loongson Technology Corporation Limited. All rights reserved.
+//
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <larchintrin.h>
+
+#include "cpuid.h"
+#include "tsc.h"
+
+#include "boot.h"
+#include "config.h"
+#include "pmem.h"
+#include "vmem.h"
+#include "memctrl.h"
+#include "memsize.h"
+#include "hwquirks.h"
+
+#include "cpuinfo.h"
+
+//------------------------------------------------------------------------------
+// Constants
+//------------------------------------------------------------------------------
+
+#define BENCH_MIN_START_ADR 0x10000000   // 256MB
+
+//------------------------------------------------------------------------------
+// Public Variables
+//------------------------------------------------------------------------------
+
+const char  *cpu_model = NULL;
+
+int         l1_cache = 0;
+int         l2_cache = 0;
+int         l3_cache = 0;
+
+uint32_t    l1_cache_speed  = 0;
+uint32_t    l2_cache_speed  = 0;
+uint32_t    l3_cache_speed  = 0;
+uint32_t    ram_speed = 0;
+
+uint32_t    clks_per_msec = 0;
+
+//------------------------------------------------------------------------------
+// Private Functions
+//------------------------------------------------------------------------------
+
+static void determine_cache_size()
+{
+  uint32_t cache_info = 0;
+  uint32_t cache_present = 0;
+  uint32_t ways = 0;
+  uint32_t index = 0;
+  uint32_t line_size = 0;
+
+  // Get the persentence of caches
+  cache_present = __cpucfg(0x10);
+
+  // L1 D cache persent
+  if (cache_present & (0x1 << 0)) {
+    cache_info = __cpucfg(0x12);
+    ways = (cache_info & 0xFFFF) + 1;
+    index = 1 << ((cache_info >> 16) & 0xFF);
+    line_size = 1 << ((cache_info >> 24) & 0x7F);
+    l1_cache = (ways * index * line_size) / 1024;
+  }
+
+  // L2 cache persent
+  if (cache_present & (0x1 << 3)) {
+    cache_info = __cpucfg(0x13);
+    ways = (cache_info & 0xFFFF) + 1;
+    index = 1 << ((cache_info >> 16) & 0xFF);
+    line_size = 1 << ((cache_info >> 24) & 0x7F);
+    l2_cache = (ways * index * line_size) / 1024;
+  }
+
+  // L3 cache persent
+  if (cache_present & (0x1 << 10)) {
+    cache_info = __cpucfg(0x14);
+    ways = (cache_info & 0xFFFF) + 1;
+    index = 1 << ((cache_info >> 16) & 0xFF);
+    line_size = 1 << ((cache_info >> 24) & 0x7F);
+    l3_cache = (ways * index * line_size) / 1024;
+  }
+}
+
+static void determine_imc(void)
+{
+  // Check Loongson IMC
+  if (cpuid_info.vendor_id.str[0] == 'L') {
```

Comment:
Do we want to check more bytes here?

Reply:
Currently, LoongArch vendor is only Loongson, I guess it's not too late for us to add more byte checks here if another vendor name also starts with the letter "L" in the future. :)